rfc/plugins.md

```diff
@@ -0,0 +1,343 @@
+
+We plan to prototype alternative options for Deck, such as altering webpack to support dynamic resolution, or introducing a standard container that can build Deck on startup with the configured plugins from Front50.
+
+### Protobuf as the Remote Contract
+As a departure from PF4J, plugin contracts will be defined as Protobuf messages (not services), so long as they are _remote-capable_. 
+If a plugin is in-process only, and there’s no expectation that it becomes available for remote plugins, it does not need to define its contract as protobuf.
+
+However, for any remotely-capable extension, its contract must be defined as protobufs, rather than exposed as simple POJOs or interfaces. 
+A key distinction is that we’re not specifying gRPC as the transport.  
+A remote transport may be Amazon SQS or Redis PubSub: not necessarily gRPC.
+
+### Technology Compatibility Kits
+_alias: TCKs_
+
+A convention of writing TCKs for each Extension Point will be established. 
+The concept of a TCK is borrowed from the JSR world, where compatibility tests are written which assert correct behavior of various features.  
+Each implementation must them pass those compatibility tests.  
+For Spinnaker plugins, TCKs can be written for two different consumers: Extension Point providers and Plugin developers.
+
+TCKs assert correct behavior of an extension point and its associated plugin(s). 
+For any extension point, a TCK will be written that asserts even pathological plugins behave in an expected way while integrating with the extension point. 
+For a plugin, a TCK will be provided that a plugin developer can use to assert correctness without necessarily needing to run other Spinnaker services.
+
+This goes further than the convention of using Protobuf to define the interfaces of a particular plugin, where Protobufs only define the interface, whereas a TCK will assert behavior that lies beneath the interface. 
+Combining a TCK and adherence to the interface, contracts will be fully asserted between a plugin and its extension point(s).
+
+### Up/Downgrading Plugins
+This section applies to federated development scenarios, where the operator is not necessarily the same as the plugin developers; where the deploy cadence of a plugin does not adhere to the deploy cadence of the core Spinnaker services.
+
+The upgrading (or downgrading, from now on broadly upgrading) plugin story is different based on the particular plugin runtime.
+
+#### In-process
+All in-process plugins should be downloadable. 
+When a service starts up, it will ask Front50 for a manifest of plugins, downloading all installed plugins that are not already present on the filesystem (in the case of baked-in plugins) to a locally-accessible directory. 
+At this point, PF4J will scan the directory, installing them into the process and enabling them based on the information pulled from Front50. 
+Should Front50 be unavailable, the service will fail to start.
```

Comment:
> Should Front50 be unavailable, the service will fail to start.

This is new territory for Spinnaker, AFAIK. Currently every service can start up on its own and just has reduced functionality until its dependencies can be contacted. This is especially helpful when you're boostrapping Spinnaker for the first time. Is this really something we're willing to give up?

As far as I can tell, the main point of having front50 as the central repository is that it means operators can see all the installed and enabled plugins. What if instead the services pushed this info into front50 at startup? This seems feasible to me, especially since we don't support changing your plugins at service runtime...

Reply:
Frankly, if we're going to do this, why not just throw away halyard and store all the config in `front50`

Reply:
Throwing away halyard is a great idea! 🤣

If, let's say, a cloud provider were a plugin, would we actually want to allow clouddriver to start? That's not reduced functionality, it's just straight up broken.

Would it be preferential from an operator perspective to install a plugin into each service, rather than services asking front50 what plugins a service needs to install?

Reply:
Ha, yes, I think we're generally in agreement there.

It seems like the three pieces of config for each plugin are (id+plugin_location+enabled). I think you're proposing a model where that information is stored in front50. It's certainly not a terrible idea on its face, but its definitely different from the status quo, where all config is stored in yaml files within each service. I'm not convinced that plugin config is so fundamentally different from other types of config that it warrants splitting the config in two.

But more importantly (and less flippantly) I just feel like moving configuration data into front50 is a pretty big change that should probably happen in a separate discussion... there are likely other things we'd consider moving and other solutions we might want to explore (Spring Cloud Config Server?).

Reply:
On the front50 hard dependency issue: What if we treated the config similar to Spring PropertySources, where if front50 is unavailable, it would fallback to config shipped with each individual service (and subsequently, services would report in what plugins + versions they're running)? This would allow us to ensure that critical plugin functionality would always be present even if front50 is unavailable, but still be able to allow plugin developers to iterate and release plugins without requiring operators to manage them.

One of the use cases that we're trying to address is that plugins are not necessarily owned by the people who are operating Spinnaker. If config is forced to live in each service, we as operators would need to be involved in any plugin changes, which would pretty well defeat a large purpose of the plugin work.

Spring Config Server is a potentially viable solution, but you're right, it would require additional thought.

The key takeaway is that we as Spinnaker operators should not be required to be in the game of managing plugin configuration. It's entirely possible that many Spinnaker deployments would have the operators and plugin developers be the same people, but in a large organization, this would not be the case.

Reply:
But the operator still has to restart any service when a new plugin is added, a plugin is removed, or a plugin version is changed, right?

> There are no current plans for supporting a world where in-process plugins can be upgraded at runtime.

I don't get how you achieve an operator-free upgrade scenario when that is the case... What use cases are you addressing here?

Reply:
We were planning to have Spinnaker redeploy itself (rolling push? not sure) when a new plugin is either installed or updated.

Reply:
I guess I don't really understand the difference, then. With front50, you push the plugin info to `front50` then restart Spinnaker. With a yaml config, you rebuild the Spinnaker container with new YAML then restart Spinnaker with that new container. I guess you don't want devs to rebuild Spinnaker containers?

Can't you just use a `latest` tag in the plugin config (e.g. `netflix.com:clouddriver-nflx:latest`) and then you don't need to do a config change for an upgrade anyway?

Reply:
The new-install story would still involve a yaml change, but do you really want people flinging new plugins into Spinnaker without any awareness from the ops team? Seems weird to me, but maybe you do...

Reply:
Personally I like idea of resolvers and letting an operator opt-in/out of the `front50` dependency (or some other plugin registry).

We (Netflix) have some flexibility around build/bake time configuration, instance launch and process launch ... I'm not sure I know enough right now to commit to one direction.

I'd like to see us move forward with some experiments, I don't see any decisions at this stage being irreversible.

Reply:
Sure, total agreement with that.

A resolver that knew how to source from `front50` would be configurable, just as one that could read from local config or an s3 bucket, etc.

Reply:
:100: I'm working on refactoring the code to have resolvers, the first and only implementation will source all plugin config from the Spring `Environment` PropertySource.